referenceframe/frame_system.go

```diff
@@ -371,6 +371,117 @@ func (sfs *simpleFrameSystem) composeTransforms(frame Frame, inputMap map[string
 	return q, errAll
 }
 
+// StartPositions returns a zeroed input map ensuring all frames have inputs.
+func StartPositions(fs FrameSystem) map[string][]Input {
+	positions := make(map[string][]Input)
+	for _, fn := range fs.FrameNames() {
+		frame := fs.Frame(fn)
+		if frame != nil {
+			positions[fn] = make([]Input, len(frame.DoF()))
+		}
+	}
+	return positions
+}
+
+// ToProtobuf turns all the interfaces into serializable types.
+func (part *FrameSystemPart) ToProtobuf() (*pb.FrameSystemConfig, error) {
+	if part.FrameConfig == nil {
+		return nil, ErrNoModelInformation
+	}
+	linkFrame, err := LinkInFrameToTransformProtobuf(part.FrameConfig)
+	if err != nil {
+		return nil, err
+	}
+	var modelJSON map[string]interface{}
+	if part.ModelFrame != nil {
+		bytes, err := part.ModelFrame.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(bytes, &modelJSON)
+		if err != nil {
+			return nil, err
+		}
+	}
+	kinematics, err := protoutils.StructToStructPb(modelJSON)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.FrameSystemConfig{
+		Frame:      linkFrame,
+		Kinematics: kinematics,
+	}, nil
+}
+
+// ProtobufToFrameSystemPart takes a protobuf object and transforms it into a FrameSystemPart.
+func ProtobufToFrameSystemPart(fsc *pb.FrameSystemConfig) (*FrameSystemPart, error) {
+	frameConfig, err := LinkInFrameFromTransformProtobuf(fsc.Frame)
+	if err != nil {
+		return nil, err
+	}
+	part := &FrameSystemPart{
+		FrameConfig: frameConfig,
+	}
+
+	if len(fsc.Kinematics.AsMap()) > 0 {
+		modelBytes, err := json.Marshal(fsc.Kinematics.AsMap())
+		if err != nil {
+			return nil, err
+		}
+		modelFrame, err := UnmarshalModelJSON(modelBytes, frameConfig.Name())
```

Comment:
this can just assign to `part.modelFrame`

Reply:
You want to check the error first though before possibly assigned an errant value to the struct. I'd keep as is